clang/test/C/C2y/n3353.c

```diff
@@ -0,0 +1,120 @@
+
+// Note that 0o by itself is an invalid literal.
+int m = 0o; /* expected-error {{invalid suffix 'o' on integer constant}}
+               c2y-warning {{octal literals without a '0o' prefix are deprecated}}
+             */
+
+// Ensure negation works as expected.
+static_assert(-0o1234 == -668); /* ext-warning {{octal integer literals are a C2y extension}}
+                                   cpp-warning {{octal integer literals are a Clang extension}}
+                                   compat-warning {{octal integer literals are incompatible with standards before C2y}}
+                                 */
+
+// FIXME: it would be better to not diagnose the compat and ext warnings when
+// the octal literal is invalid.
+// We expect diagnostics for non-octal digits.
+int n = 0o18; /* expected-error {{invalid digit '8' in octal constant}}
+                 compat-warning {{octal integer literals are incompatible with standards before C2y}}
+                 ext-warning {{octal integer literals are a C2y extension}}
+                 cpp-warning {{octal integer literals are a Clang extension}}
+               */
+int o1 = 0o8; /* expected-error {{invalid suffix 'o8' on integer constant}}
+                 c2y-warning {{octal literals without a '0o' prefix are deprecated}}
+               */
+// FIXME: however, it matches the behavior for hex literals in terms of the
+// error reported. Unfortunately, we then go on to think 0 is an octal literal
+// without a prefix, which is again a bit confusing.
+int o2 = 0xG; /* expected-error {{invalid suffix 'xG' on integer constant}}
+                 c2y-warning {{octal literals without a '0o' prefix are deprecated}}
+               */
+
+// Ensure digit separators work as expected.
+constexpr int p = 0o0'1'2'3'4'5'6'7; /* compat-warning {{octal integer literals are incompatible with standards before C2y}}
+                                        ext-warning {{octal integer literals are a C2y extension}}
+                                        cpp-warning {{octal integer literals are a Clang extension}}
+                                      */
+static_assert(p == 01234567); // c2y-warning {{octal literals without a '0o' prefix are deprecated}}
+int q = 0o'0'1; /* expected-error {{invalid suffix 'o'0'1' on integer constant}}
+                   c2y-warning {{octal literals without a '0o' prefix are deprecated}}
+                 */
+
+#define M 0o123
+int r = M;  /* compat-warning {{octal integer literals are incompatible with standards before C2y}}
+               ext-warning {{octal integer literals are a C2y extension}}
+               cpp-warning {{octal integer literals are a Clang extension}}
+             */
+
+// Also, test delimited escape sequences. Note, this paper added a delimited
+// escape sequence for octal *and* hex.
+auto a = "\x{12}\o{12}";   /* compat-warning 2 {{delimited escape sequences are incompatible with C standards before C2y}}
```

Comment:
```suggestion
auto a = "\x{12}\o{12}\N{SPARKLES}";   /* compat-warning 2 {{delimited escape sequences are incompatible with C standards before C2y}}
```

Reply:
We should have a **clang** extension warning in all C modes (it's a C++-only feature that we support as an extension in C)

Reply:
Yes, we should, but what does this have to do with octal and hexadecimal escape sequences? ;-) I'll add it as a drive-by.